google/auth/pluggable.py

```diff
@@ -0,0 +1,327 @@
+            subject_token_type=subject_token_type,
+            token_url=token_url,
+            credential_source=credential_source,
+            service_account_impersonation_url=service_account_impersonation_url,
+            client_id=client_id,
+            client_secret=client_secret,
+            quota_project_id=quota_project_id,
+            scopes=scopes,
+            default_scopes=default_scopes,
+            workforce_pool_user_project=workforce_pool_user_project,
+        )
+        if not isinstance(credential_source, Mapping):
+            self._credential_source_executable = None
+            raise ValueError(
+                "Missing credential_source. The credential_source is not a dict."
+            )
+        else:
+            self._credential_source_executable = credential_source.get("executable")
+            if not self._credential_source_executable:
+                raise ValueError(
+                    "Missing credential_source. An 'executable' must be provided."
+                )
+            self._credential_source_executable_command = self._credential_source_executable.get(
+                "command"
+            )
+            self._credential_source_executable_timeout_millis = self._credential_source_executable.get(
+                "timeout_millis"
+            )
+            self._credential_source_executable_output_file = self._credential_source_executable.get(
+                "output_file"
+            )
+
+            # environment_id is only supported in AWS or dedicated future external
+            # account credentials.
+            if "environment_id" in credential_source:
+                raise ValueError(
+                    "Invalid Pluggable credential_source field 'environment_id'"
+                )
+
+        if not self._credential_source_executable_command:
+            raise ValueError("Missing command. Executable command must be provided.")
+        if not self._credential_source_executable_timeout_millis:
+            raise ValueError(
+                "Missing timeout_millis. Executable timeout millis must be provided."
+            )
+
+    @_helpers.copy_docstring(external_account.Credentials)
+    def retrieve_subject_token(self, request):
+        env_allow_executables = os.environ.get(
+            "GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES"
+        )
+        if env_allow_executables != "1":
+            raise ValueError(
+                "Executables need to be explicitly allowed (set GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES to '1') to run."
+            )
+
+        # Check output file
+        if self._credential_source_executable_output_file is not None:
+            try:
+                with open(
+                    self._credential_source_executable_output_file
+                ) as output_file:
+                    response = json.load(output_file)
+                    subject_token = self._parse_subject_token(response)
+            except:
+                pass
```

Comment:
I think ignoring issues when trying to read from the file hides bugs and maybe we should just throw an error. WDYT?

Reply:
The output file is more of a cache here. So if the file is broken, we just invalidate the cache and retrieve the token again. Throwing an error requires the developer to manually correct the output file, which might be hard.

Reply:
If the file is corrupted, or acl'd badly, or something else, then we are hiding this from the developer and it is not behaving as intended. I think throwing an error here instead ensures that their configuration is properly set up and everything is WAI.

Reply:
Make sense.

Reply:
There are multiple cases that we should rerun the executable. For example: the file is not found, the token is expired, etc.. So we can't just throw an exception here.
How about we rerun the executable, but print in the console to inform the developer that the cached output file is not valid?

Reply:
Done.

Reply:
Seems like there is missing a change here? Still just ignoring any issues with the output file?

Reply:
If there is an issue with the file itself (not the content) like file not exists, the error is ignored and we generate the file then.
If the there is an issue with the content, the error is raised.